src/table_provider.rs

```diff
@@ -0,0 +1,2007 @@
+use std::collections::HashSet;
+use async_trait::async_trait;
+use zarrs::filesystem::FilesystemStore;
+use zarrs::group::Group;
+use zarrs::array::Array;
+use zarrs::array::data_type::DataType as ZarrDataType;
+use zarrs::array_subset::ArraySubset;
+use zarrs::array::chunk_grid::ChunkGrid;
+use arrow_array::{Int16Array, Int64Array, Float64Array, Int32Array, Float32Array};
+
+/// Represents a coordinate range constraint for chunk filtering
+#[derive(Debug, Clone)]
+struct CoordinateRange {
+    dimension: usize,
+    min: Option<i64>,
+    max: Option<i64>,
+}
+
+impl CoordinateRange {
+    fn new(dimension: usize) -> Self {
+        Self { dimension, min: None, max: None }
+    }
+    
+    fn with_min(mut self, min: i64) -> Self {
+        self.min = Some(min);
+        self
+    }
+    
+    fn with_max(mut self, max: i64) -> Self {
+        self.max = Some(max);
+        self
+    }
+    
+    /// Check if a coordinate value satisfies this range
+    fn contains(&self, value: i64) -> bool {
+        if let Some(min) = self.min {
+            if value < min { return false; }
+        }
+        if let Some(max) = self.max {
+            if value > max { return false; }
+        }
+        true
+    }
+}
+
+/// Represents a collection of coordinate constraints for filtering
+#[derive(Debug, Clone)]
+pub struct CoordinateFilter {
+    ranges: Vec<CoordinateRange>,
+}
+
+impl CoordinateFilter {
+    fn new() -> Self {
+        Self { ranges: Vec::new() }
+    }
+    
+    fn add_range(mut self, range: CoordinateRange) -> Self {
+        self.ranges.push(range);
+        self
+    }
+    
+    /// Check if a set of coordinates satisfies all constraints
+    pub fn matches(&self, coordinates: &[i64]) -> bool {
+        for range in &self.ranges {
+            if range.dimension < coordinates.len() {
+                if !range.contains(coordinates[range.dimension]) {
+                    return false;
+                }
+            }
+        }
+        true
+    }
+}
+
+/// Trait for types that can be converted to Arrow arrays with minimal copying
+trait ToArrowArray: Clone + Sized {
+    type ArrowArray: arrow_array::Array + 'static;
+    
+    /// Convert a flattened ndarray to an Arrow array with zero-copy when possible
+    fn to_arrow_array(
+        flat_data: &ndarray::ArrayBase<ndarray::CowRepr<'_, Self>, ndarray::Dim<[usize; 1]>>,
+    ) -> Vec<Self>;
+    
+    /// Get the Arrow DataType for this type
+    fn arrow_data_type() -> DataType;
+    
+    /// Create an Arrow array from a vector
+    fn from_vec(data: Vec<Self>) -> Arc<Self::ArrowArray>;
+}
+
+impl ToArrowArray for f64 {
+    type ArrowArray = Float64Array;
+    
+    fn to_arrow_array(
+        flat_data: &ndarray::ArrayBase<ndarray::CowRepr<'_, Self>, ndarray::Dim<[usize; 1]>>,
+    ) -> Vec<Self> {
+        if flat_data.is_standard_layout() {
+            flat_data.as_slice().unwrap().to_vec()
+        } else {
+            flat_data.iter().cloned().collect()
+        }
```

Comment:
fyi I think this will never be zero-copy. Essentially any time you're calling `to_vec` on a slice it'll copy it to a new Vec